File: app/src/main/java/com/app/messagealarm/model/dao/ApplicationDao.java
package com.app.messagealarm.model.dao;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.Query;
import androidx.room.Update;

import com.app.messagealarm.model.entity.ApplicationEntity;
import com.app.messagealarm.model.response.AppTypes;

import java.util.ArrayList;
import java.util.List;

@Dao
public interface ApplicationDao {
    @Insert
    void insertApplication(ApplicationEntity applicationEntity);

    @Query("SELECT * FROM applications")
    List<ApplicationEntity> getAllApplicationList();

    @Update
    int updateApplication(ApplicationEntity applicationEntity);

    @Query("UPDATE applications SET running_status = :status WHERE id = :id")
    void updateAppStatus(boolean status, int id);

    @Query("UPDATE applications SET vibrate_on_alarm = :status")
    void disableVibrateToAllApp(boolean status);

    @Query("UPDATE applications SET just_vibrate = :status")
    void disableJustVibrateToAllApp(boolean status);

    @Query("DELETE FROM applications")
    void deleteAllAppsWithoutTheFirstOne();

    @Query("UPDATE applications SET sound_level = :sound_level")
    void rollBackAppsFromDefaultSoundLevel(int sound_level);

    @Delete
    void deleteApplication(ApplicationEntity applicationEntity);

    @Query("SELECT * FROM applications WHERE package_name = :packageName")
    ApplicationEntity getAppByPackageName(String packageName);

    @Query("SELECT * FROM applications WHERE package_name = :packageName AND is_alarm_enabled = :isAlarmEnabled")
    ApplicationEntity getAppByPackageNameAndAlarm(String packageName, Boolean isAlarmEnabled);

    @Query("SELECT * FROM applications WHERE package_name = :packageName AND is_speak_enabled = :isSpeakEnabled")
    ApplicationEntity getAppByPackageNameAndSpeak(String packageName, Boolean isSpeakEnabled);

    @Query("SELECT package_name, MAX(is_alarm_enabled = 'true') as isAlarmConfigured, " +
            "MAX(is_speak_enabled = 'true') as isSpeakConfigured" +
            " FROM applications WHERE package_name = :packageName GROUP BY package_name")
    AppTypes getAppTypesByPackage(String packageName);

    @Query("SELECT COUNT(*) FROM applications")
    int getAddedAppCount();

}

File: app/src/main/java/com/app/messagealarm/model/dao/ApplicationDao.kt
package com.app.messagealarm.model.dao

import androidx.lifecycle.LiveData
import androidx.room.*
import com.app.messagealarm.model.entity.ApplicationEntity
import com.app.messagealarm.model.entity.ApplicationTable

@Dao
interface ApplicationDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(application: ApplicationEntity)

    @Update
    fun update(application: ApplicationEntity)

    @Delete
    fun delete(application: ApplicationEntity)

    @Query("UPDATE applications SET running_status = :status WHERE id = :id")
    fun updateAppStatus(status: Boolean, id: Int)

    @Query("SELECT * FROM ${ApplicationTable.TABLE_NAME}")
    fun allApplications(): LiveData<List<ApplicationEntity>>

    @Query("SELECT * FROM ${ApplicationTable.TABLE_NAME}")
    fun getAllApplications(): List<ApplicationEntity>

    @Query("SELECT * FROM ${ApplicationTable.TABLE_NAME} WHERE ${ApplicationTable.ID} = :id")
    fun getApplication(id: Int): ApplicationEntity

    @Query("SELECT * FROM ${ApplicationTable.TABLE_NAME} WHERE ${ApplicationTable.PACKAGE_NAME} = :packageName")
    fun getAppByPackageName(packageName: String): ApplicationEntity

    @Query("DELETE FROM ${ApplicationTable.TABLE_NAME}")
    fun deleteAllApplications()
}
